src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts

```diff
@@ -360,26 +375,51 @@ export class DataView implements IIndexPattern {
 
   /**
    * Add a runtime field - Appended to existing mapped field or a new field is
-   * created as appropriate
+   * created as appropriate. We can pass along with the fields a "format" definition or "customLabel" to avoid calling
+   * separately `setFieldCustomLabel()` and `setFieldFormat()`.
    * @param name Field name
-   * @param runtimeField Runtime field definition
+   * @param enhancedRuntimeField Runtime field definition
    */
-  addRuntimeField(name: string, runtimeField: RuntimeField) {
+  addRuntimeField(name: string, enhancedRuntimeField: EnhancedRuntimeField): DataViewField {
+    const existRuntimeCompositeWithSameName = this.getRuntimeComposite(name) !== null;
+
+    if (existRuntimeCompositeWithSameName) {
+      throw new Error(
+        `Can't add runtime field ["${name}"] as there is already a runtime composite with the same name.`
+      );
+    }
+
+    const { type, script, parentComposite, customLabel, format, popularity } = enhancedRuntimeField;
+
+    const runtimeField: RuntimeField = { type, script, parentComposite };
+    this.runtimeFieldMap[name] = runtimeField;
```

Comment:
It looks like as a developer of these services I could easily misuse the functions on the class. For example, if I call

`addScriptedField('foo'...)` and then `addRuntimeField('foo'` I think I'm going to end up with a `DataViewField` that has `this.spec.runtimeField` _and_ `this.script` set at the same time.

FWIW, that doesn't look new in this PR.

Reply:
Great catch. Yes indeed we will need to add the same validation as I did on L386. But that's not new to this PR. I can add it in a future work.